File: services/stream-processor/influx.go
package main

import (
	"context"
	"fmt"
	"time"

	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
	"github.com/influxdata/influxdb-client-go/v2/api"
	"github.com/influxdata/influxdb-client-go/v2/api/write"
	"github.com/nithya-prakash/indusense/pkg/reliability"
)

// influxSink wraps the InfluxDB client with retry-with-backoff and a circuit
// breaker, matching the same pattern used for Kafka writes in ingestion —
// InfluxDB is just as much an external dependency that can be temporarily
// unavailable, and deserves the same fail-fast-during-an-outage treatment.
type influxSink struct {
	client   influxdb2.Client
	writeAPI api.WriteAPIBlocking
	org      string
	bucket   string

	breaker    *reliability.CircuitBreaker
	maxRetries int
	retryDelay time.Duration
}

func newInfluxSink(cfg Config) *influxSink {
	client := influxdb2.NewClient(cfg.InfluxURL, cfg.InfluxToken)
	return &influxSink{
		client:     client,
		writeAPI:   client.WriteAPIBlocking(cfg.InfluxOrg, cfg.InfluxBucket),
		org:        cfg.InfluxOrg,
		bucket:     cfg.InfluxBucket,
		breaker:    reliability.NewCircuitBreaker(cfg.BreakerFailureThreshold, cfg.BreakerCooldown),
		maxRetries: cfg.InfluxMaxRetries,
		retryDelay: cfg.InfluxRetryBaseDelay,
	}
}

func (s *influxSink) close() {
	s.client.Close()
}

func (s *influxSink) ping(ctx context.Context) error {
	_, err := s.client.Ping(ctx)
	return err
}

func (s *influxSink) breakerState() string {
	return s.breaker.State()
}

// writeRawPoint writes one sensor_telemetry point. Measurement/tags/fields
// match docs/DATABASE.md exactly: tags identify the hierarchy path + metric,
// the only field is value. Writing the same (tags, timestamp) twice is a
// safe no-op overwrite, which is what makes this naturally idempotent under
// at-least-once delivery.
func (s *influxSink) writeRawPoint(ctx context.Context, key seriesKey, at time.Time, value float64) error {
	p := write.NewPoint("sensor_telemetry",
		map[string]string{
			"factory_id":         key.FactoryID,
			"production_line_id": key.ProductionLineID,
			"machine_id":         key.MachineID,
			"device_id":          key.DeviceID,
			"sensor_id":          key.SensorID,
			"metric":             key.Metric,
		},
		map[string]any{"value": value},
		at,
	)
	return s.writeWithProtection(ctx, p)
}

// buildAggregatePoint constructs (without writing) one sensor_telemetry_agg
// point for a given window (e.g. "1m"), for batching by the caller.
func buildAggregatePoint(key seriesKey, window string, at time.Time, stats windowStats) *write.Point {
	return write.NewPoint("sensor_telemetry_agg",
		map[string]string{
			"factory_id":         key.FactoryID,
			"production_line_id": key.ProductionLineID,
			"machine_id":         key.MachineID,
			"device_id":          key.DeviceID,
			"sensor_id":          key.SensorID,
			"metric":             key.Metric,
			"window":             window,
		},
		map[string]any{
			"moving_avg":     stats.MovingAvg,
			"moving_stddev":  stats.MovingStdDev,
			"min":            stats.Min,
			"max":            stats.Max,
			"rate_of_change": stats.RateOfChange,
			"count":          stats.Count,
		},
		at,
	)
}

func (s *influxSink) writeWithProtection(ctx context.Context, points ...*write.Point) error {
	if !s.breaker.Allow() {
		return fmt.Errorf("circuit breaker open for influxdb")
	}

	err := reliability.RetryWithBackoff(ctx, s.maxRetries, s.retryDelay, func(d time.Duration) { time.Sleep(d) }, func() error {
		return s.writeAPI.WritePoint(ctx, points...)
	})

	if err != nil {
		s.breaker.RecordFailure()
		return err
	}
	s.breaker.RecordSuccess()
	return nil
}

// writeAggregateBatch writes many aggregate points (one per series x window,
// for a single flush tick) as a single InfluxDB call instead of one round
// trip per point — with up to ~1000 sensors x 5 windows per flush, that
// difference matters.
func (s *influxSink) writeAggregateBatch(ctx context.Context, points []*write.Point) error {
	if len(points) == 0 {
		return nil
	}
	return s.writeWithProtection(ctx, points...)
}

File: services/stream-processor/influx.py
"""InfluxDB writes with retry + circuit breaker, the Python port of
influx.go. Wraps the InfluxDB client with retry-with-backoff and a circuit
breaker, matching the same pattern used for Kafka writes in ingestion —
InfluxDB is just as much an external dependency that can be temporarily
unavailable, and deserves the same fail-fast-during-an-outage treatment.
"""

from __future__ import annotations

from datetime import datetime

from influxdb_client import InfluxDBClient, Point
from influxdb_client.client.write_api import SYNCHRONOUS

from config import Config
from registry import SeriesKey
from shared.reliability import CircuitBreaker, retry_with_backoff
from window import WindowStats


class InfluxSink:
    def __init__(self, cfg: Config):
        self._client = InfluxDBClient(url=cfg.influx_url, token=cfg.influx_token, org=cfg.influx_org)
        self._write_api = self._client.write_api(write_options=SYNCHRONOUS)
        self._org = cfg.influx_org
        self._bucket = cfg.influx_bucket
        self._breaker = CircuitBreaker(cfg.breaker_failure_threshold, cfg.breaker_cooldown_seconds)
        self._max_retries = cfg.influx_max_retries
        self._retry_delay = cfg.influx_retry_base_delay_seconds

    def close(self) -> None:
        self._client.close()

    def ping(self) -> bool:
        return self._client.ping()

    def breaker_state(self) -> str:
        return self._breaker.state()

    def write_raw_point(self, key: SeriesKey, at: datetime, value: float) -> None:
        """Writes one sensor_telemetry point. Measurement/tags/fields match
        docs/DATABASE.md exactly: tags identify the hierarchy path + metric,
        the only field is value. Writing the same (tags, timestamp) twice is
        a safe no-op overwrite, which is what makes this naturally
        idempotent under at-least-once delivery."""
        point = (
            Point("sensor_telemetry")
            .tag("factory_id", key.factory_id)
            .tag("production_line_id", key.production_line_id)
            .tag("machine_id", key.machine_id)
            .tag("device_id", key.device_id)
            .tag("sensor_id", key.sensor_id)
            .tag("metric", key.metric)
            .field("value", value)
            .time(at)
        )
        self._write_with_protection(point)

    def write_aggregate_batch(self, points: list[Point]) -> None:
        """Writes many aggregate points (one per series x window, for a
        single flush tick) as a single InfluxDB call instead of one round
        trip per point — with up to ~1000 sensors x 5 windows per flush,
        that difference matters."""
        if not points:
            return
        self._write_with_protection(*points)

    def _write_with_protection(self, *points: Point) -> None:
        if not self._breaker.allow():
            raise RuntimeError("circuit breaker open for influxdb")

        try:
            retry_with_backoff(
                self._max_retries,
                self._retry_delay,
                lambda: self._write_api.write(bucket=self._bucket, org=self._org, record=list(points)),
            )
        except Exception:
            self._breaker.record_failure()
            raise
        self._breaker.record_success()


def build_aggregate_point(key: SeriesKey, window_label: str, at: datetime, stats: WindowStats) -> Point:
    """Constructs (without writing) one sensor_telemetry_agg point for a
    given window (e.g. "1m"), for batching by the caller."""
    return (
        Point("sensor_telemetry_agg")
        .tag("factory_id", key.factory_id)
        .tag("production_line_id", key.production_line_id)
        .tag("machine_id", key.machine_id)
        .tag("device_id", key.device_id)
        .tag("sensor_id", key.sensor_id)
        .tag("metric", key.metric)
        .tag("window", window_label)
        .field("moving_avg", stats.moving_avg)
        .field("moving_stddev", stats.moving_stddev)
        .field("min", stats.min)
        .field("max", stats.max)
        .field("rate_of_change", stats.rate_of_change)
        .field("count", stats.count)
        .time(at)
    )
